### PR TITLE
docs: add initial performance guide doc

### PR DESCRIPTION
**Pull request by borkmann**

### CODE REVIEW

File: Documentation/spelling_wordlist.txt [hunk 1]
```diff
@@ -684,6 +688,7 @@ uprobes
 upstreamed
 uretprobes
 url
+usecs
```
Comment (joestringer) [inline Documentation/spelling_wordlist.txt]: nit: I didn't see another mention of this in the doc, maybe not necessary?

File: Documentation/concepts/performance/index.rst [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+**********************
+
+This document presents the steps to evaluate Cilium networking performance.
+Our primary goal is to make the evaluation
+process transparent and its results easily reproducible. In this document, we
+focus on small-scale experiments. For larger deployments, refer to the :ref:`scalability_guide`.
+
+To ease reproducibility, this report is paired with a set of scripts that can be
+found in https://github.com/cilium/perfeval (all scripts in this document refer
+to this repository). Specifically, we use
```
Comment (joestringer) [inline Documentation/concepts/performance/index.rst]: nit, brackets not necessary.
```suggestion
To ease reproducibility, this report is paired with a set of scripts that can be
found in https://github.com/cilium/perfeval. All scripts in this document refer
to this repository. Specifically, we use
```

File: Documentation/concepts/performance/index.rst [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+`Terraform <https://www.terraform.io/>`_ and `Ansible
+<https://www.ansible.com/>`_ to setup the environment and execute benchmarks.
+We use `Packet <https://www.packet.com/>`_ bare metal servers as our hardware
+platform, but the guide is structured so that it can be easily adapted to
+others.
```
Comment (joestringer) [inline Documentation/concepts/performance/index.rst]:
```suggestion
other environments.
```

File: Documentation/concepts/performance/index.rst [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+
+
+The above will provision two servers named ``knb-0`` and ``knb-1`` of type
+``c3.small.x86`` and configure them to use a "Mixed/Hybrid" network mode under a
+common VLAN named ``knb``.  The machines will be provisioned with an
+``ubuntu_20_04`` OS.  We also create a ``packet-hosts.ini`` file to use as an
+inventory file for Ansible.
+
+Verify that everything works as expected by executing an ad-hoc ``uptime``
+command on the servers.
```
Comment (joestringer) [inline Documentation/concepts/performance/index.rst]:
```suggestion
Verify that the servers are successfully provisioned by executing an ad-hoc ``uptime``
command on the servers.
```

File: Documentation/concepts/performance/index.rst [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+Install ``kubeadm`` and its dependencies:
+
+.. code-block:: shell-session
+
+  $ ansible-playbook -i packet-hosts.ini playbooks/install-kubeadm.yaml
+
+We use `kubenetbench <https://github.com/cilium/kubenetbench>`_ to execute the
+`netperf <https://hewlettpackard.github.io/netperf/>`_ benchmark in a Kubernetes
+environment.  While kubenetbench was developed by Cilium developers, it is not
+specific to Cilium and can be executed on top of any Kubernetes deployment. In
```
Comment (joestringer) [inline Documentation/concepts/performance/index.rst]: nit: This statement sounds a bit defensive. We are placing these tools out into the public in a spirit of transparency, so the fact that some Cilium developers were involved should not matter. If someone is concerned about that, they can open issues on that repository to discuss how to adjust for any potential bias.

Perhaps:
```suggestion
environment.  kubenetbench is a Kubernetes benchmarking project that is agnostic
to the CNI or networking plugin that the cluster is deployed with. In
```

or 

```suggestion
environment. In
```

File: Documentation/concepts/performance/index.rst [hunk 6]
```diff
@@ -0,0 +1,330 @@
...
+---------
+
+Configure Cilium in tunneling (:ref:`arch_overlay`) mode:
+
+.. code-block:: shell-session
+
+  $ ansible-playbook -e mode=tunneling -i packet-hosts.ini playbooks/install-k8s-cilium.yaml
+  $ ansible-playbook -e conf=vxlan -i packet-hosts.ini playbooks/run-kubenetbench.yaml
+
+The first command configures Cilium to use tunneling (``-e mode=tunneling``),
```
Comment (joestringer) [inline Documentation/concepts/performance/index.rst:171]: Do we need to add any additional commands in between the above to ensure that the machines are restarted & ready? Maybe it should be split into two code blocks to make sure that anyone trying to reproduce will stop to verify this first?

File: Documentation/concepts/performance/index.rst [hunk 7]
```diff
@@ -0,0 +1,330 @@
...
+  :width: 450
+  :alt: TCP maerts
+
+Request/Reply Messages
+----------------------
+
+Next, we examine the performance of sending small (1 byte) request and reply
+messages between a client and a server. Even though many studies focus on
+bandwidth measurements, modern applications rely heavily on message passing and
+this benchmark captures their behavior better.
```
Comment (joestringer) [inline Documentation/concepts/performance/index.rst]: nit:

```suggestion
this benchmark captures their behavior more accurately.
```

**Review by joestringer — CHANGES_REQUESTED** [review]
I have a scattering of minor wording fixups below, but LGTM overall.

We should add "higher is better" / "lower is better" wording to every graph. The results may not be so apparent to people who are less familiar with reading these graphs.

File: Documentation/concepts/performance/index.rst (hunk 6 above)
Comment (borkmann) [reply Documentation/concepts/performance/index.rst:171]: Will leave this one to potentially address by either @pchaigno or @kkourt .

Comment (borkmann) [reply Documentation/concepts/performance/index.rst:171]: @kkourt could you also follow-up to include the "higher is better" / "lower is better" wording into every plot?

Comment (pchaigno) [reply Documentation/concepts/performance/index.rst:171]: Why would we need to restart the machines? If we need to, we can do it as part of the playbook.

File: Documentation/spelling_wordlist.txt (hunk 1 above)
Comment (pchaigno) [reply Documentation/spelling_wordlist.txt]: Oops, I removed that word from the doc.

File: Documentation/concepts/performance/index.rst (hunk 6 above)
Comment (joestringer) [reply Documentation/concepts/performance/index.rst:171]: Sorry, I meant to add this feedback for line 216 which talks about clearing out old state.

Comment (pchaigno) [reply Documentation/concepts/performance/index.rst:171]: Ah I see. Maybe we could just check that BPF programs are unloaded as part of the second playbook (`playbooks/run-rawnetperf.yaml`). That should just be a matter of running `bpftool net`. I'm guessing that would be our main concern, right?

**Review by joestringer — APPROVED** [review]

**Comment by kkourt** [comment]
> "higher is better" / "lower is better"

Good point! I'll add proper wording.

File: Documentation/concepts/performance/index.rst (hunk 6 above)
Comment (kkourt) [reply Documentation/concepts/performance/index.rst:171]: AFAIU, the ansible play will wait until the machine _is_ rebooted so state should be clean after that. There are two reasons for the reboot: bpf programs (for which we could call cilium cleanup and check bpf progs), but also changes on the MTU of the NICs.

**Comment by borkmann** [comment]
(I'll readd the 1.8 backport once also the follow-ups are merged early next week so we can take in all at once w/o partial state.)

**Pull request merged**